libcnb-test/src/lib.rs

```diff
@@ -0,0 +1,260 @@
+    tokio_runtime: tokio::runtime::Runtime,
+}
+
+/// References a Cloud Native Buildpack
+pub enum BuildpackReference {
+    /// References this buildpack in the Rust Crate under test
+    Crate,
+    /// References another buildpack by id, local directory or tarball
+    Other(String),
+}
+
+impl IntegrationTest {
+    /// Creates a new integration test.
+    ///
+    /// # Panics
+    /// - When the connection to Docker failed
+    /// - When the internal Tokio runtime could not be created
+    pub fn new(builder_name: impl Into<String>, app_dir: impl AsRef<Path>) -> Self {
+        IntegrationTest {
+            app_dir: PathBuf::from(app_dir.as_ref()),
+            target_triple: String::from("x86_64-unknown-linux-musl"),
+            builder_name: builder_name.into(),
+            buildpacks: vec![BuildpackReference::Crate],
+            docker: Docker::connect_with_local_defaults()
+                .expect("Could not connect to local Docker deamon"),
+            tokio_runtime: tokio::runtime::Runtime::new()
+                .expect("Could not create internal Tokio runtime"),
+        }
+    }
+
+    /// Sets the buildpacks order.
+    ///
+    /// Defaults to [`BuildpackReference::Crate`].
+    pub fn buildpacks(&mut self, buildpacks: impl Into<Vec<BuildpackReference>>) -> &mut Self {
+        self.buildpacks = buildpacks.into();
+        self
+    }
+
+    /// Sets the target triple.
+    ///
+    /// Defaults to `x86_64-unknown-linux-musl`.
+    pub fn target_triple(&mut self, target_triple: impl Into<String>) -> &mut Self {
+        self.target_triple = target_triple.into();
+        self
+    }
+
+    /// Starts a new integration test run.
+    ///
+    /// This function will copy the application to a temporary directory, cross-compiles this crate,
+    /// packages it as a buildpack and then invokes [pack](https://buildpacks.io/docs/tools/pack/)
+    /// to build a new Docker image with the buildpacks specified by this integration test instance.
+    ///
+    /// Since this function is supposed to only be used in integration tests, failures are not
+    /// signalled via [`Result`](Result) values. Instead, this function panics whenever an unexpected error
+    /// occurred to simplify testing code.
+    ///
+    /// # Panics
+    /// - When the app could not be copied
+    /// - When this crate could not be packed as a buildpack
+    /// - When the "pack" command unexpectedly fails
+    ///
+    /// # Example
+    /// ```no_run
+    /// use libcnb_test::IntegrationTest;
+    ///
+    /// IntegrationTest::new("heroku/buildpacks:20", "test-fixtures/app")
+    ///     .run(|context| {
+    ///         assert!(context.pack_stdout.contains("---> Ruby Buildpack"));
+    ///         assert!(context.pack_stdout.contains("---> Installing bundler"));
+    ///         assert!(context.pack_stdout.contains("---> Installing gems"));
+    ///     })
+    /// ```
+    pub fn run<F: FnOnce(IntegrationTestContext)>(&mut self, f: F) {
```

Comment:
Lastly, if I modify `examples/example-02-ruby-sample/buildpack.toml` such that it doesn't match the spec (eg rename `stacks` to `foo`), I get the following:

```
$ cargo test
...
     Running tests/integration_test.rs (/Users/emorley/src/libcnb.rs/target/debug/deps/integration_test-1bb9860726cd382d)

running 1 test
    Finished dev [unoptimized + debuginfo] target(s) in 0.50s
test basic ... FAILED

failures:

---- basic stdout ----
thread 'basic' panicked at 'assertion failed: context.pack_stdout.contains(\"---> Ruby Buildpack\")', examples/example-02-ruby-sample/tests/integration_test.rs:11:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    basic

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 3.23s

error: test failed, to rerun pass '--test integration_test'
```

In contrast, when I run `cargo libcnb package`, I get:

```
$ cargo libcnb package
INFO - Reading buildpack metadata...
ERROR - Unable to deserialize buildpack metadata: unknown field `foo`, expected one of `api`, `buildpack`, `stacks`, `metadata` at line 11 column 1
ERROR - Hint: Verify that your "buildpack.toml" is valid.
```

Reply:
Almost all of these are caused by one root-cause: The error output would be in the `pack` output but it is captured for matching against it in the tests. If the tests in the Ruby example would use `assert_eq` the error output would've been shown.

I opened an issue for handling missing `pack`: 
- https://github.com/Malax/libcnb.rs/issues/296

> In contrast, when I run cargo libcnb package, I get

This is because `cargo-libcnb` reads the `buildpack.toml`, but `libcnb-test` doesn't. It's a coincidence this error scenario has handling outside of `pack` ouput. Any other error would've lead to the

Reply:
Possible failure modes (mainly noting these for my benefit):
1. Docker not installed (could detect via `which` crate or `bollard`'s [Docker::ping()](https://docs.rs/bollard/0.11.1/bollard/struct.Docker.html#method.ping))
2. Docker daemon not running (could detect via `bollard`'s [Docker::ping()](https://docs.rs/bollard/0.11.1/bollard/struct.Docker.html#method.ping))
3. Pack not installed (could detect via `which` crate)
4. Pack installed, but a version that doesn't support the current buildpack API (could detect via scraping `pack version`, but bleugh, perhaps best to just fall back on displaying pack stderr?)
5. Buildpack does not compile successfully (already handled via this PR)
6. Pack build fails due to invalid `buildpack.toml` (could read the `buildpack.toml` in libcnb-test like libcnb-cargo does, or else just fall back on displaying pack stderr)

Reply:
So I think libcnb-test needs to:
1. By default treat a failing `pack build` as a test failure (currently it does not, the test will only fail if one of the assertions inside the test also happens to fail, which may not always be the case)
2. When a test fails due to (1), output the stderr/stdout to aid debugging
3. Provide a way to specify a test as "allowed to fail", to opt out of (1)

Solving (1) I think is a must-have before this merges.

Solving (1)+(2) will also happen to improve the UX for the other failure modes listed above.

Reply:
> currently it does not, the test will only fail if one of the assertions inside the test also happens to fail

For example, if I change `examples/example-02-ruby-sample/tests/integration_test.rs` such that the `basic` test only does an `assert!(true);` inside the function passed to `IntegrationTest::run()`, then the test passes.

Now clearly `assert!(true);` is not representative of a real test, but imagine a test that tries to confirm an EOL runtime message **is not** shown:

```rust
assert!(!context.pack_stdout.contains("Security update available"));
```

This would still pass even if the build failed.

Reply:
💯  agree that we, when `pack` exists with non-zero, should panic and therefore fail the test. I'll implement that.

Reply:
Implemented in 6f6402a, example panic:
```
test basic ... FAILED

failures:

---- basic stdout ----
thread 'basic' panicked at 'pack command failed with exit-code 1!

pack stdout:


pack stderr:
ERROR: failed to build: failed to fetch builder image 'index.docker.io/heroku/buildpocks:20': image 'index.docker.io/heroku/buildpocks:20' does not exist on the daemon: not found
', /Users/manuel.fuchs/projects/libcnb.rs/libcnb-test/src/lib.rs:189:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    basic

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.26s

error: test failed, to rerun pass '--test integration_test'```

Reply:
Thank you for implementing that!

I think we might want a new GitHub issue for:
1. Adding support for tests that are allowed to fail (eg so we can write a test that checks detection fails when appropriate)
2. Adding integration tests to the libcnb-test crate itself, which can test things like this change (since currently libcnb-test is only really integration tested via the examples, and we probably don't want a `#[should_panic]` test in our examples to test 6f6402a etc).

Reply:
See dbd0214 for alternative _2_. I think _1_ is out of scope for this concrete issue and would require more thought before implementation.

Reply:
Thank you for adding an integration test for point (2).

> I think 1 is out of scope for this concrete issue 

That's why I said we should file a GitHub issue for it :-)

Reply:
> Adding support for tests that are allowed to fail (eg so we can write a test that checks detection fails when appropriate)

Filed #300 for this.

Reply:
> If I run `cargo test` in the root of the libcnb repo, without the Docker daemon running, the error message says:

New error message for this, after 6f6402a:

```
test basic ... FAILED

failures:

---- basic stdout ----
thread 'basic' panicked at 'pack command failed with exit-code 1!

pack stdout:


pack stderr:
ERROR: failed to build: failed to fetch builder image 'index.docker.io/heroku/buildpacks:20': Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
', /Users/emorley/src/libcnb.rs/libcnb-test/src/lib.rs:198:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    basic

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.75s
```

> Lastly, if I modify examples/example-02-ruby-sample/buildpack.toml such that it doesn't match the spec (eg rename stacks to foo

New error message for this, after 6f6402a:

```
test basic ... FAILED

failures:

---- basic stdout ----
thread 'basic' panicked at 'pack command failed with exit-code 1!

pack stdout:
20: Pulling from heroku/buildpacks
Digest: sha256:b6165ec47a7d25046570887e7929a5b411495477a8f822a52b3987055618e53d
Status: Image is up to date for heroku/buildpacks:20
20: Pulling from heroku/pack
Digest: sha256:b7de85bb2dd228a5bee87813c1a6c61fd81668e4abe7b1a77a7ead66d59ab35b
Status: Image is up to date for heroku/pack:20


pack stderr:
ERROR: failed to build: downloading buildpack: extracting from 'file:///var/folders/sl/fwxwgr4d0pncnv36t27w9dxr0000gq/T/.tmpwMAC0C': reading buildpack: invalid buildpack.toml: buildpack 'libcnb-examples/ruby@0.1.0': must have either 'stacks' or an 'order' defined
', /Users/emorley/src/libcnb.rs/libcnb-test/src/lib.rs:198:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    basic

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 3.79s
```

These look great! The remaining cases are tracked at #296 and #300, so we can call this thread resolved for this PR :-)